### ISSUE TITLE
Problems with pagination in v1.1.8

### ISSUE THREAD

**Issue by manuelmeurer** [mailgun/mailgun-ruby#124]
I recently updated to v1.1.8 of this gem (from v1.1.6) and my code started to fail.
I tracked it down to the pagination that seems to be buggy.

What I'm doing is getting the "delivered" events for two of my domains:

```
mailgun = Mailgun::Client.new
%w(domain1.com domain2.com).each do |domain|
  mailgun_events = Mailgun::Events.new(mailgun, domain)
  events = mailgun_events.get(event: 'delivered').to_h['items']
  while events.any?
    events.each do |event|
      # Do something with the event
    end
    events = mailgun_events.next.to_h['items']
  end
end
```

What started happening is that not only "delivered", but ALL events were returned from the API.
When I downgraded to v1.1.6 again, the error went away.

I assume the bug is somewhere in here: https://github.com/mailgun/mailgun-ruby/commit/9ad5cbd95f2570b3a3c681cc1e362c702b84eded#diff-e904dd8e3437a9a9c5e35c6dcb227d30

**Comment by manuelmeurer** [comment]
Can I do anything to help debug this further?

**Comment by jrosengren** [comment]
I'm seeing similar behavior in 1.1.9.

**Comment by scottrobertson** [comment]
```
while(result = events.next) do
  break if result.to_h['items'].size == 0
  store_events(result.to_h['items'])
end
```

This just loops for ever on this version too

**Comment by scottrobertson** [comment]
When calling `.next` it returns the last page over and over again. 

```
    events = Mailgun::Events.new(client, domain)
    result = events.get(
      begin:      1.year.ago.to_i,
      end:        Time.now.to_i,
      limit:      100,
      ascending:  'yes',
      event:      'delivered OR opened OR clicked'
    )
```

`result` is from 1 year ago,

`events.next` now returns the latest 100 events

**Comment by scottrobertson** [comment]
It's also ignoring my `limit`

**Comment by aldrienht** [comment]
encounters the same issue in pagination :(

**Comment by fursich** [comment]
Just realized the discussion going on here - I still haven't been able to dig each of the issues raised above, but as @manuelmeurer pointed out, I come to suspect the uri parser is not doing the right job (huge thanks to the detailed reports!)

My assumption is, it's ignoring query strings dubbed at the end of the page URIs each time when jumping to other pages.

Let me dig it further down and submit a PR for the fix right away.

**Comment by fursich** [comment]
Although I wasn't able to identify the reason why it had originally worked in the past versions, but at least a quick analysis helped me understand what happened with the above cases.

## TL; DR

mailgun's event API 'forgets' parameters EACH TIME you fetch the data.

So I'd suggest to set parameters when calling `next` and `previous`, as it now [accepts arguments](https://github.com/mailgun/mailgun-ruby/commit/364ecda594be8ab9b8a1930f6350c6e59f598904).

For instance, in your case @aldrienht , instead of simply hitting `next`, `previous`

```ruby
@all_mails = @mg_events.next.to_h['items']
```

you could ** specify the parameters with every request **

```ruby
event_settings = { limit: 5. pretty: 'yes', event: 'delivered' }

@all_mails = @mg_events.next(event_settings).to_h['items']
```

As far as I can see the mailgun-ruby gem hasn't been changed its behavior about how to handle the parameters in recent versions, so this would be the only way to keep the original settings by far.

Hope that solves the problems above.

FYI, suppressions API seems to be working differently. It returns requested params each time as part of the next url, which allows us to retrieve bounced mails with the same conditions over again. (you simply can hit `next` and `previous` without specifying parameters)

## API responses

Below shows part of what I have done for an analysis.

You could quickly check the API feature as follows:

```ruby
# requesting two mails directly (without using the gem)

response = RestClient.get( "https://api:YOUR_API_KEY@api.mailgun.net/v3/YOUR_DOMAIN/events", params: { limit: 2 } )

JSON.parse(response.body)['paging']['next']
=> https://api.mailgun.net/v3/YOUR_DOMAIN/events/YACI4pW3siLTA0LTA.....

# params are not set as part of the next URI
```

On the other hand, when you request bounces..

```ruby
# requesting two mails directly (without using the gem)

response = RestClient.get( "https://api:YOUR_API_KEY@api.mailgun.net/v3/YOUR_DOMAIN/events", params: { limit: 2 } )

JSON.parse(response.body)['paging']['next']
=> "https://api.mailgun.net/v3/YOUR_DOMAIN/bounces?page=next&address=somebody%40example.com&limit=2"

# params are embedded as part of the next URI
```

Due to this feature, you are requested to set parameters (eg. limit, begin, end..etc) each time, or otherwise mailgun seems to forget the past settings, and responds based on its default settings.

I'm not really sure why/how it worked in the past, but it apparently shows it's coming from API feature itself.

Hope that saves somebody's day..

**Comment by fursich** [comment]
It might be a good idea, though, to add a new feature with this gem to recall event params once given, but it might add a distractive change on the Mailgun::Event's behavior. (What would you think, @pirogoeth ?)

**Comment by manuelmeurer** [comment]
@fursich Thanks for digging in! In your second example where you're requesting bounces, the URL should be `...YOUR_DOMAIN/bounces` instead of `...YOUR_DOMAIN/events` though I think.

Have you contacted Mailgun about this? Seems like a bug that the response from `/events` is different from `/bounces`.

I agree that this gem should remember the params which were used on the initial get... it wouldn't make sense to change them when calling `next`.

**Comment by aldrienht** [comment]
@fursich Yes it works in `next` but no result when hitting `previous` a bit weird.
But I did the same thing in `previous`:

   `other_params = { limit: 1, pretty: 'yes' }`
   `# Want more results?`
   ` if params[:next].present?`
      ` @all_mails = @mg_events.next(other_params.merge({ event: 'delivered' })).to_h['items']`
   ` end`

   `# Go backwards?`
   ` if params[:previous].present?`
      ` @all_mails = @mg_events.previous(other_params.merge({ event: 'delivered' })).to_h['items']`
   ` end`

**Comment by fursich** [comment]
@manuelmeurer 
Thanks for correction! Updated my comment above.

Yes, considering the nature of this wrapper gem, I personally think it's reasonable to remember the params. Am happy to volunteer but as it involves a major changes in terms of how Event responds, I'd like to hear how this gem's committers think. (I'm just a very occasional contributor)

Haven't contacted Mailgun but would be great if any of you could volunteer to!
I assume it might not be going to happen until next V4 versions to come (as it affects all the users), but definitely worth raising our voices 👍 

@aldrienht 

I have no idea on that, but the previous page might not be present as you are on the first page.
Have you checked the "raw" API response I showed above?  If the raw response contains some mails and you get no data via mailgun-ruby, then it's probably a new (different) issue.

FYI, it seems that Mailgun always responds with a url - even when you are at the first page and the suggested "previous" url contains zero mails you always get something. Check out if you have more than two pages of emails, otherwise you are always on the page where "previous link" makes no sense.

**Comment by manuelmeurer** [comment]
I opened a ticket for the Mailgun support about this and will keep you updated about their response!

**Comment by manuelmeurer** [comment]
Answer from Mailgun: `I made some API calls to the the Events API and the Next URL returned appears to be working correctly.  I think this may be specifically related to the Ruby SDK.  Are you able to extract the full API call and responses that the Ruby API calls are making?`
So this seems to be a problem with the way this gem calls the API or parses/returns the `next` and `previous` URLs.

**Comment by kaspergrubbe** [comment]
So this bit me in the foot with version 1.1.9, however none of the solutions seems to work.

**Comment by kaspergrubbe** [comment]
I have been digging a bit deeper, it seems like the regex in `Mailgun::Events.extract_endpoint_from` is broken, https://github.com/mailgun/mailgun-ruby/blob/32f9d729321ffda475ab69c3ce310ba42ff7cb17/lib/mailgun/events/events.rb#L97-L107

If I feed it next-page-urls that the API is returning, it always returns `nil`.

**Comment by kaspergrubbe** [comment]
I think I have solved the problem here: https://github.com/mailgun/mailgun-ruby/pull/142

**Comment by n-rodriguez** [comment]
Hi there! Still not working here.

I have 140_000 events to retrieve and nothing works.

![screenshot_2018-09-18 logs list - mailgun](https://user-images.githubusercontent.com/3433835/45658097-73f12880-baed-11e8-90f1-6e0a2101303f.png)

I've tested everything (passing params to `next` or `previous` methods), hacking the lib, etc... and I'm a bit desperate to make it work.

I confirm that the lib doesn't remember passed params. On the top of that, I don't know how the  `each` method could work since you can't pass search params to it (sic).

**Comment by n-rodriguez** [comment]
I thought this would be sufficient to retrieve all the events of that day, but no :

```ruby
  params = {
    'begin' => Time.parse('2018-09-08 00:00:00 UTC').utc.rfc2822,
    'end'   => Time.parse('2018-09-09 00:00:00 UTC').utc.rfc2822,
    'limit' => 300,
  }

  loop do
    result = events.next(params)
    result = result.to_h
    parse_data(result)
    break if result['items'].empty?
  end
```

The [Time Range](https://documentation.mailgun.com/en/latest/api-events.html#time-range) thing is not very clear. How does it behave when you don't pass `begin` and `end`? What if I only search by subject? 

```ruby
  params = {
    'subject' => 'Foo',
    'limit' => 300,
  }

  loop do
    result = events.next(params)
    result = result.to_h
    parse_data(result)
    break if result['items'].empty?
  end
```

**Comment by Retttro** [comment]
Fixed in https://github.com/mailgun/mailgun-ruby/pull/142 and published to rubygems. Closing.
